### PR TITLE
Revert Shift Anti-Drag and Disable in PvP Scenarios

### PR DESCRIPTION
**Pull request by jcwhisman**
The changes made reflect the exact same sentiments stated by @dekvall  [in PR #10870](https://github.com/runelite/runelite/pull/10870).

The following features have been reverted/added to the anti-drag plugin:
- **Shift Anti-Drag** plugin has been reverted back to original **Anti Drag**
- _Shift Only_ anti-drag still remains an option.
- Any form of anti-drag is disabled in PvP scenarios. (Detected by the spec orb enabling/disabling). Including in PvP worlds and wilderness.
- Anti drag is enabled for banking.

With the addition of removing all forms of anti-drag for PvP scenarios, I believe this sufficiently solves the issue/uproar in the PvP community that caused the plugin to get reverted in the first place.

Example on a PvP world:
rDJcg9x

### CODE REVIEW

**Comment by jcwhisman** [comment]
I know there are still some discussions on how runelite should handle anti drag. I just haven't received any feedback on this pr yet. Where are the admins at in those discussions?

**Comment by Adam-** [comment]
I think this approach of just disabling it in pvp via the specorb varbit is the way we want to go, but I haven't looked at this yet.

File: runelite-client/src/main/java/net/runelite/client/plugins/antidrag/AntiDragPlugin.java [hunk 1]
```diff
@@ -65,13 +77,27 @@ AntiDragConfig getConfig(ConfigManager configManager)
...
+		}
+
 		keyManager.registerKeyListener(this);
 	}
 
 	@Override
 	protected void shutDown() throws Exception
 	{
 		client.setInventoryDragDelay(DEFAULT_DELAY);
+		setBankDragDelay(DEFAULT_DELAY);
```
Comment (Adam-) [inline runelite-client/src/main/java/net/runelite/client/plugins/antidrag/AntiDragPlugin.java]: This actually looks like a bug fix? I may take this independently/it should be in a separate commit.

File: runelite-client/src/main/java/net/runelite/client/plugins/antidrag/AntiDragConfig.java [hunk 2]
```diff
@@ -40,6 +40,24 @@
...
-		return Constants.GAME_TICK_LENGTH / Constants.CLIENT_TICK_LENGTH; // one game tick
+		return Constants.GAME_TICK_LENGTH / Constants.CLIENT_TICK_LENGTH;
+	}
+
+	@ConfigItem(
+		keyName = "dragDelay",
+		name = "",
+		description = ""
+	)
+	void dragDelay(int delay);
```
Comment (Adam-) [inline runelite-client/src/main/java/net/runelite/client/plugins/antidrag/AntiDragConfig.java]: I can't figure out what this is here for. I don't think it's being used?

File: runelite-client/src/main/java/net/runelite/client/plugins/antidrag/AntiDragPlugin.java [hunk 3]
```diff
@@ -24,32 +24,44 @@
...
-	description = "Prevent dragging an item for a specified delay",
-	tags = {"antidrag", "delay", "inventory", "items"}
+	name = "Anti Drag",
+	enabledByDefault = false
 )
+@Slf4j
 public class AntiDragPlugin extends Plugin implements KeyListener
 {
-	private static final int DEFAULT_DELAY = 5;
+	static final String CONFIG_GROUP = "antiDrag";
```
[the remark this answers is not in the document]
Comment (jcwhisman) [reply runelite-client/src/main/java/net/runelite/client/plugins/antidrag/AntiDragPlugin.java]: I'm sorry I am still trying to learn my way around github and understand the review process. Are you referring to line 55 - `static final String CONFIG_GROUP = "antiDrag";`? If so, are you suggesting referencing the `ConfigGroup`
via `AntiDragConifg.class.getAnnotation(ConfigGroup.class).value()`?

[the remark this answers is not in the document]
Comment (Nightfirecat) [reply runelite-client/src/main/java/net/runelite/client/plugins/antidrag/AntiDragPlugin.java]: He means that `AntiDragConfig` should also have its `@ConfigGroup` definition updated to reference this constant. So it would read like this:

```java
@ConfigGroup(AntiDragPlugin.CONFIG_GROUP)
```

File: runelite-client/src/main/java/net/runelite/client/plugins/antidrag/AntiDragPlugin.java (hunk 1 above)
Comment (jcwhisman) [reply runelite-client/src/main/java/net/runelite/client/plugins/antidrag/AntiDragPlugin.java]: Referring to the shutdown behavior? This really is there because the original anti-drag plugin had it. I will test to validate.

File: runelite-client/src/main/java/net/runelite/client/plugins/antidrag/AntiDragPlugin.java [hunk 4]
```diff
@@ -125,4 +193,9 @@ private void setBankDragDelay(int delay)
 		}
 	}
 
+	public boolean inPvp()
+	{
+		return client.getVarbitValue(8121) == 1;
```
[the remark this answers is not in the document]
Comment (jcwhisman) [reply runelite-client/src/main/java/net/runelite/client/plugins/antidrag/AntiDragPlugin.java]: Can I keep the Varbits.java file change and make this `client.getVar(Varbits.PVP_SPEC_ORB_DISABLED)` or is it better practice to not modify the varbits and just use `client.getVarbitValue(8121)`?

File: runelite-client/src/main/java/net/runelite/client/plugins/antidrag/AntiDragPlugin.java [hunk 5]
```diff
@@ -84,32 +110,74 @@ public void keyTyped(KeyEvent e)
...
 		}
 	}
 
 	@Subscribe
-	public void onFocusChanged(FocusChanged focusChanged)
+	public void onVarbitChanged(VarbitChanged varbitChanged)
 	{
-		if (!focusChanged.isFocused())
+
+		if (!inPvp() && !config.onShiftOnly())
```
[the remark this answers is not in the document]
Comment (jcwhisman) [reply runelite-client/src/main/java/net/runelite/client/plugins/antidrag/AntiDragPlugin.java]: Are you referring to setting isPvp to false in onFocusChanged?

File: runelite-client/src/main/java/net/runelite/client/plugins/antidrag/AntiDragPlugin.java [hunk 6]
```diff
@@ -24,32 +24,45 @@
...
 import net.runelite.client.input.KeyListener;
 import net.runelite.client.input.KeyManager;
 import net.runelite.client.plugins.Plugin;
 import net.runelite.client.plugins.PluginDescriptor;
 
+
 @PluginDescriptor(
-	name = "Shift Anti Drag",
-	description = "Prevent dragging an item for a specified delay",
-	tags = {"antidrag", "delay", "inventory", "items"}
```
Comment (Nightfirecat) [inline runelite-client/src/main/java/net/runelite/client/plugins/antidrag/AntiDragPlugin.java]: Why are the plugin description and tags removed?

File: runelite-client/src/main/java/net/runelite/client/plugins/antidrag/AntiDragPlugin.java [hunk 7]
```diff
@@ -24,32 +24,45 @@
...
+	name = "Anti Drag",
+	enabledByDefault = false
 )
+@Slf4j
```
Comment (Nightfirecat) [inline runelite-client/src/main/java/net/runelite/client/plugins/antidrag/AntiDragPlugin.java]: You're not using any `log` statements... Remove this.

File: runelite-client/src/main/java/net/runelite/client/plugins/antidrag/AntiDragPlugin.java [hunk 8]
```diff
@@ -24,32 +24,45 @@
...
 public class AntiDragPlugin extends Plugin implements KeyListener
 {
-	private static final int DEFAULT_DELAY = 5;
+	static final String CONFIG_GROUP = "antiDrag";
+
+	static final int DEFAULT_DELAY = 5;
```
Comment (Nightfirecat) [inline runelite-client/src/main/java/net/runelite/client/plugins/antidrag/AntiDragPlugin.java]: Is this being used outside the plugin? Seems like it can remain `private`.

File: runelite-client/src/main/java/net/runelite/client/plugins/antidrag/AntiDragPlugin.java [hunk 9]
```diff
@@ -84,32 +111,76 @@ public void keyTyped(KeyEvent e)
...
+			clientThread.invoke(() ->
+			{
+				client.setInventoryDragDelay(DEFAULT_DELAY);
+				setBankDragDelay(DEFAULT_DELAY);
+			});
 		}
 	}
 
 	@Subscribe
-	public void onFocusChanged(FocusChanged focusChanged)
```
Comment (Nightfirecat) [inline runelite-client/src/main/java/net/runelite/client/plugins/antidrag/AntiDragPlugin.java]: Why is the `FocusChanged` handler removed?

File: runelite-client/src/main/java/net/runelite/client/plugins/antidrag/AntiDragPlugin.java [hunk 10]
```diff
@@ -84,32 +111,76 @@ public void keyTyped(KeyEvent e)
...
+				if (!config.onShiftOnly() && !inPvp())
+				{
+					client.setInventoryDragDelay(config.dragDelay());
+					setBankDragDelay(config.dragDelay());
+				}
+				else
+				{
+					client.setInventoryDragDelay(DEFAULT_DELAY);
+					setBankDragDelay(DEFAULT_DELAY);
+				}
```
Comment (Nightfirecat) [inline runelite-client/src/main/java/net/runelite/client/plugins/antidrag/AntiDragPlugin.java]: These blocks are duplicated numerous times in this plugin now. Move them to helper methods.

File: runelite-api/src/main/java/net/runelite/api/Varbits.java [hunk 11]
```diff
@@ -563,7 +563,12 @@
 	TWISTED_LEAGUE_RELIC_2(10050),
 	TWISTED_LEAGUE_RELIC_3(10051),
 	TWISTED_LEAGUE_RELIC_4(10052),
-	TWISTED_LEAGUE_RELIC_5(10053);
+	TWISTED_LEAGUE_RELIC_5(10053),
+
+	/**
+	 * PVP
```
Comment (Nightfirecat) [inline runelite-api/src/main/java/net/runelite/api/Varbits.java]: This should explain the expected function and values of this varbit, not just be a section header of sorts. (Not all other entries in this file are a good reference; I would expect this to look more like `GE_OFFER_CREATION_TYPE` on L542 than the sections above)

File: runelite-client/src/main/java/net/runelite/client/plugins/antidrag/AntiDragConfig.java (hunk 2 above)
Comment (jcwhisman) [reply runelite-client/src/main/java/net/runelite/client/plugins/antidrag/AntiDragConfig.java]: This was an accident. I have removed it.

File: runelite-client/src/main/java/net/runelite/client/plugins/antidrag/AntiDragPlugin.java (hunk 9 above)
Comment (jcwhisman) [reply runelite-client/src/main/java/net/runelite/client/plugins/antidrag/AntiDragPlugin.java]: I removed it at first because I didn't understand why it was important to turn off delay when the client left focus. I have added it back.

File: runelite-client/src/main/java/net/runelite/client/plugins/antidrag/AntiDragPlugin.java [hunk 12]
```diff
@@ -48,14 +48,15 @@
...
+	description = "Prevent dragging an item for a specified delay",
+	tags = {"antidrag", "delay", "inventory", "items"}
 )
-@Slf4j
 public class AntiDragPlugin extends Plugin implements KeyListener
 {
 	static final String CONFIG_GROUP = "antiDrag";
 
-	static final int DEFAULT_DELAY = 5;
+	private final int DEFAULT_DELAY = 5;
```
Comment (Nightfirecat) [inline runelite-client/src/main/java/net/runelite/client/plugins/antidrag/AntiDragPlugin.java]: This can remain `static` in addition to being private.
```suggestion
	private static final int DEFAULT_DELAY = 5;
```

File: runelite-api/src/main/java/net/runelite/api/Varbits.java [hunk 13]
```diff
@@ -563,7 +563,14 @@
 	TWISTED_LEAGUE_RELIC_2(10050),
 	TWISTED_LEAGUE_RELIC_3(10051),
 	TWISTED_LEAGUE_RELIC_4(10052),
-	TWISTED_LEAGUE_RELIC_5(10053);
+	TWISTED_LEAGUE_RELIC_5(10053),
+
+	/**
+	 * Status of the PvP Spec Orb
+	 * 0 = Disabled
+	 * 1 = Enabled
```
Comment (Nightfirecat) [inline runelite-api/src/main/java/net/runelite/api/Varbits.java]: This javadoc does not match the implementation of `inPvp()`. I would expect it to read more like the following:
```suggestion
	 * Whether the Special Attack orb is disabled due to being in a PvP area
	 *
	 * 0 = Enabled (player is not in PvP)
	 * 1 = Disabled (player in in PvP)
	 *
	 * @see https://oldschool.runescape.wiki/w/Minimap#Special_attack_orb
```

File: runelite-client/src/main/java/net/runelite/client/plugins/antidrag/AntiDragConfig.java (hunk 2 above)
Comment (Nightfirecat) [reply runelite-client/src/main/java/net/runelite/client/plugins/antidrag/AntiDragConfig.java]: This still has not been removed. Why was this comment marked resolved?

File: runelite-client/src/main/java/net/runelite/client/plugins/antidrag/AntiDragPlugin.java [hunk 14]
```diff
@@ -24,54 +24,82 @@
...
+	@Inject
+	private ClientThread clientThread;
+
 	@Inject
 	private AntiDragConfig config;
 
 	@Inject
 	private KeyManager keyManager;
 
+	private boolean lastStatus; //true = was in PvP, false = was not in Pvp
```
Comment (Nightfirecat) [inline runelite-client/src/main/java/net/runelite/client/plugins/antidrag/AntiDragPlugin.java]: `lastInPvp` would be a much better var name, which would completely remove the need for this added comment.

File: runelite-client/src/main/java/net/runelite/client/plugins/antidrag/AntiDragConfig.java (hunk 2 above)
Comment (jcwhisman) [reply runelite-client/src/main/java/net/runelite/client/plugins/antidrag/AntiDragConfig.java]: Sorry. I did remove this error, but someone accidentally re-pushed it back out.

File: runelite-api/src/main/java/net/runelite/api/Varbits.java [hunk 15]
```diff
@@ -563,7 +563,17 @@
...
-	TWISTED_LEAGUE_RELIC_5(10053);
+	TWISTED_LEAGUE_RELIC_5(10053),
+
+	/**
+	 * Whether the Special Attack orb is disabled due to being in a PvP area
+	 *
+	 * 0 = Enabled (player is not in PvP)
+	 * 1 = Disabled (player in in PvP)
+	 *
+	 * @see https://oldschool.runescape.wiki/w/Minimap#Special_attack_orb
```
Comment (Nightfirecat) [inline runelite-api/src/main/java/net/runelite/api/Varbits.java]: Turns out my suggestion wasn't proper javadoc. This should be though.

```suggestion
	 * @see <a href="https://oldschool.runescape.wiki/w/Minimap#Special_attack_orb">The OSRS Wiki's Minimap page</a>
```

File: runelite-client/src/main/java/net/runelite/client/plugins/antidrag/AntiDragPlugin.java [hunk 16]
```diff
@@ -84,31 +110,86 @@ public void keyTyped(KeyEvent e)
 	@Override
 	public void keyPressed(KeyEvent e)
 	{
-		if (e.getKeyCode() == KeyEvent.VK_SHIFT)
+		if (config.onShiftOnly() && e.getKeyCode() == KeyEvent.VK_SHIFT)
```
Comment (Adam-) [inline runelite-client/src/main/java/net/runelite/client/plugins/antidrag/AntiDragPlugin.java:113]:
```suggestion
		if (e.getKeyCode() == KeyEvent.VK_SHIFT && config.onShiftOnly())
```

File: runelite-client/src/main/java/net/runelite/client/plugins/antidrag/AntiDragPlugin.java [hunk 17]
```diff
@@ -84,31 +110,86 @@ public void keyTyped(KeyEvent e)
...
-			setBankDragDelay(delay);
+			setDragDelay();
 		}
 	}
 
 	@Override
 	public void keyReleased(KeyEvent e)
 	{
-		if (e.getKeyCode() == KeyEvent.VK_SHIFT)
+		if (config.onShiftOnly() && e.getKeyCode() == KeyEvent.VK_SHIFT)
```
Comment (Adam-) [inline runelite-client/src/main/java/net/runelite/client/plugins/antidrag/AntiDragPlugin.java:122]:
```suggestion
		if (e.getKeyCode() == KeyEvent.VK_SHIFT && config.onShiftOnly())
```

File: runelite-client/src/main/java/net/runelite/client/plugins/antidrag/AntiDragPlugin.java [hunk 18]
```diff
@@ -65,13 +80,24 @@ AntiDragConfig getConfig(ConfigManager configManager)
...
+		}
+
 		keyManager.registerKeyListener(this);
 	}
 
 	@Override
 	protected void shutDown() throws Exception
 	{
-		client.setInventoryDragDelay(DEFAULT_DELAY);
+		resetDragDelay();
```
Comment (Adam-) [inline runelite-client/src/main/java/net/runelite/client/plugins/antidrag/AntiDragPlugin.java:100]: This should probably be on the client thread

File: runelite-client/src/main/java/net/runelite/client/plugins/antidrag/AntiDragPlugin.java [hunk 19]
```diff
@@ -84,31 +110,86 @@ public void keyTyped(KeyEvent e)
...
-			setBankDragDelay(DEFAULT_DELAY);
+			resetDragDelay();
+		}
+		else
+		{
+			clientThread.invokeLater(() ->
+			{
+				if (!config.onShiftOnly() && !inPvp())
+				{
+					setDragDelay();
```
Comment (Adam-) [inline runelite-client/src/main/java/net/runelite/client/plugins/antidrag/AntiDragPlugin.java:181]: Only the set (and probably the above resetDragDelay()) needs to be on client thread, the rest is fine off.

File: runelite-client/src/main/java/net/runelite/client/plugins/antidrag/AntiDragPlugin.java [hunk 20]
```diff
@@ -125,4 +206,21 @@ private void setBankDragDelay(int delay)
 		}
 	}
 
+	private boolean inPvp()
+	{
+		return client.getVar(Varbits.PVP_SPEC_ORB) == 1;
```
Comment (Adam-) [inline runelite-client/src/main/java/net/runelite/client/plugins/antidrag/AntiDragPlugin.java:211]: This is just `lastInPvp `?
